### PR TITLE
LightGraphs integration

### PR DESCRIPTION
**Pull request by GearsAD**

### CODE REVIEW

**Comment by Affie** [comment]
Failures temporary allowed on v1.2 for Metagraphs. Should be fixed in master, waiting for new release.

**Comment by Affie** [comment]
Looks like it should be MetaGraphs v0.6.4, but its not registered yet. v0.6.4 will hopefully fix build on julia 1.2

**Comment by dehann** [comment]
FYI JuliaGraphs/MetaGraphs.jl#84

File: src/LightGraphsDFG/services/LightGraphsDFG.jl [hunk 1]
```diff
@@ -0,0 +1,545 @@
...
+Get an adjacency matrix for the DFG, returned as a Matrix{Union{Nothing, Symbol}}.
+Rows are all factors, columns are all variables, and each cell contains either nothing or the symbol of the relating factor.
+The first row and first column are factor and variable headings respectively.
+"""
+function getAdjacencyMatrix(dfg::LightGraphsDFG)::Matrix{Union{Nothing, Symbol}}
+    varLabels = sort(map(v->v.label, getVariables(dfg)))
+    factLabels = sort(map(f->f.label, getFactors(dfg)))
+    vDict = Dict(varLabels .=> [1:length(varLabels)...].+1)
+
+    adjMat = Matrix{Union{Nothing, Symbol}}(nothing, length(factLabels)+1, length(varLabels)+1)
```
Comment (dehann) [inline src/LightGraphsDFG/services/LightGraphsDFG.jl]: One thing we should start doing across the entire code base is avoid type instability -- a major performance hog.  It's okay to use Union for function dispatch, thats still fast, but dynamic types require run time type inference each time.  I think this is happening in various places in the current IIF code anyway, but likelihood is high that we will run a campaign to remove all type instability from the code to improve performance.

cc @GearsAD

Comment (dehann) [reply src/LightGraphsDFG/services/LightGraphsDFG.jl]: I'm not sure if we will get type stability at the end of IIF v0.7.x (probably not) or rather do that cleanup during IIF v0.8.x cycle.  v0.8.x cannot be as unstable as IIF v0.7.x is though.  v0.7.x is bit of a hot mess.  As long as IIF 0.6.x and v0.8.x are reliable form version control standpoint.  v0.7.x is a little tricky in places.

Comment (Affie) [reply src/LightGraphsDFG/services/LightGraphsDFG.jl]: Should we change to missing? is missing and nothing treated the same? As far as I know missing has little performance impact.

**Comment by Affie** [comment]
> FYI [JuliaGraphs/MetaGraphs.jl#84](https://github.com/JuliaGraphs/MetaGraphs.jl/issues/84)

File: src/LightGraphsDFG/services/LightGraphsDFG.jl (hunk 1 above)
Comment (dehann) [reply src/LightGraphsDFG/services/LightGraphsDFG.jl]: I don't know missing that well unfortunately.  The union typedef (not function dispatch) is a bigger story.  This is part of the current DFG design so im scared if LightGraphsDFG is out of step then things are going to get weird.  My first reaction is to just update all of DFG in one go.

Comment (GearsAD) [reply src/LightGraphsDFG/services/LightGraphsDFG.jl]: In this case we need a nullable type. TBD - Let's do DFG in one go?

Comment (Affie) [reply src/LightGraphsDFG/services/LightGraphsDFG.jl]: I don't think we need to do anything in this case, julia arrays are fast with Nothing unions. 
Union{Nothing,T} = nullable type.

**Comment by GearsAD** [comment]
LGTM!

**Comment by GearsAD** [comment]
Created [IIF #368](https://github.com/JuliaRobotics/IncrementalInference.jl/issues/368) to capture effort  to support this driver for solving.

File: src/LightGraphsDFG/services/LightGraphsDFG.jl (hunk 1 above)
Comment (dehann) [reply src/LightGraphsDFG/services/LightGraphsDFG.jl]: FYI [avoid containers with abstract types](https://docs.julialang.org/en/v1/manual/performance-tips/#Avoid-containers-with-abstract-type-parameters-1)

**Pull request merged**